### PR TITLE
Add Kraken2 component 

### PR DESCRIPTION
**Pull request by cimendes**
- Adds Kraken2 with minikraken2DB 
- changed kraken, kraken2 and midas results location from `results/annotation` to `results/taxonomy`
- adds reports to maxbin2 (sample specific)
- fix issue where maxbin2 process wouldn't output binning files but would run successfully

### CODE REVIEW

File: flowcraft/templates/process_tsv.py [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+    logger.debug("SAMPLE_ID: {}".format(SAMPLE_ID))
+    logger.debug("FILE: {}".format(FILE))
+
+@MainWrapper
+def main(sample_id, tsv_file):
+
+    with open(".report.json", "w") as k:
+        k.write('{"tsvData":[{"sample":"' + sample_id + '", "data": { "MaxBin2":')
+        k.write(json.dumps(list(csv.reader(open(tsv_file), delimiter='\t'))))
+        k.write('}}]}')
```
Comment (tiagofilipe12) [inline flowcraft/templates/process_tsv.py]: I am not sure of what is going on here, but this seems a rather poor writing to the file :smirk:. Maybe you should load the content of `tsv_file` into a dict and then use `json.dumps` to dump the whole dict instead of this sketchy solution.
Basically define a dict like this:

```python
# first define a backbone of your dict like this
your_dict = {
    "tsvData": [{
        "sample": sample_id,
        "data": {}
    }]
}
# then add your entry to the dict. I am not sure why you use a list for `tsvData` since you put everything into `MaxBin2` but I hope you get the idea anyway.
your_dict["tsvData"][0]["data"]["MaxBin2"] = list(csv.reader(open(tsv_file)
# then json dump it to file like this
k.write(json.dumps(your_dict))
``` 

The whole idea is to construct a dict before dumping into the file rather than sketching a dict to a file with strings

File: docs/user/components/megahit.rst [hunk 2]
```diff
@@ -27,6 +27,10 @@ Parameters
   from the maximum read length of each assembly. If 'default', megahit will
   use the default k-mer lengths.
 
+- ``fastg``: When true, it converts megahit intermediate contigs into fastg
```
[the remark this answers is not in the document]
Comment (cimendes) [reply docs/user/components/megahit.rst]: Added. Thanks for noticing

File: docs/user/components/maxbin2.rst [hunk 3]
```diff
@@ -46,4 +46,4 @@ Default directives
 - ``container``: flowcraft/maxbin2
 - ``version``: 2.2.4-1
 - ``cpus``: 3
-- ``memory``: 5.GB * task.attempt
\ No newline at end of file
+- ``memory``: 5.GB * task.attempt
```
[the remark this answers is not in the document]
Comment (cimendes) [reply docs/user/components/maxbin2.rst]: Thanks for noticing!

File: flowcraft/generator/process_details.py [hunk 4]
```diff
@@ -68,7 +68,9 @@ def procs_dict_parser(procs_dict):
     logger.info(colored_print(
         "\n===== L I S T   O F   P R O C E S S E S =====\n", "green_bold"))
 
-    for template, dict_proc_info in procs_dict.items():
+    procs_dict_ordered = {k: procs_dict[k] for k in sorted(procs_dict)}
```
[the remark this answers is not in the document]
Comment (cimendes) [reply flowcraft/generator/process_details.py]: Added in the latest commit

File: flowcraft/generator/templates/kraken2.nf [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+
+    output:
+    file("${sample_id}_kraken_report.txt")
+    {% with task_name="kraken2" %}
+    {%- include "compiler_channels.txt" ignore missing -%}
+    {% endwith %}
+
+    script:
+    """
+    kraken2 --memory-mapping --threads $task.cpus --report ${sample_id}_kraken_report.txt --db ${krakenDB} --paired --gzip-compressed ${fastq_pair[0]} ${fastq_pair[1]}
```
[the remark this answers is not in the document]
Comment (cimendes) [reply flowcraft/generator/templates/kraken2.nf]: Addressed in the latest commit

File: flowcraft/templates/process_tsv.py [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+ to generate a report in json format.
+Expected input
+--------------
+The following variables are expected whether using NextFlow or the
+:py:func:`main` executor.
+- ``sample_id`` : Sample Identification string.
+- ``tsv``: tsv output.
+Generated output
+----------------
+- ``.report.jason``: Data structure for the report
```
[the remark this answers is not in the document]
Comment (cimendes) [reply flowcraft/templates/process_tsv.py]: :fearful:

File: flowcraft/generator/templates/maxbin2.nf [hunk 7]
```diff
@@ -50,9 +56,45 @@ process maxbin2_{{ pid }} {
...
+
+// If maxbin fails to obtain bins for a sample, the workflow continues with the original assembly
 {{ output_channel }} = Channel.create()
 
-binCh_{{ pid }}.flatMap().map{ it -> [it.toString().tokenize('/').last().tokenize('.')[0..-2].join('.'), it] }
-    .into({{ output_channel }})
+OUT_binned = Channel.create()
+OUT_unbinned = Channel.create()
+
+chanA = Channel.create()
```
[the remark this answers is not in the document]
Comment (cimendes) [reply flowcraft/generator/templates/maxbin2.nf]: Changed in the latest commit

File: docs/user/components/maxbin2.rst [hunk 8]
```diff
@@ -45,5 +45,11 @@ Default directives
 
 - ``container``: flowcraft/maxbin2
 - ``version``: 2.2.4-1
-- ``cpus``: 3
-- ``memory``: 5.GB * task.attempt
+- ``cpus``: 4
+- ``memory``: 8.GB (dynamically increased on retry)
```
Comment (tiagofilipe12) [inline docs/user/components/maxbin2.rst]: Are we now adopting this on the docs? Rather than the task.attempt nomenclature?

File: flowcraft/generator/process_details.py [hunk 9]
```diff
@@ -68,6 +68,7 @@ def procs_dict_parser(procs_dict):
     logger.info(colored_print(
         "\n===== L I S T   O F   P R O C E S S E S =====\n", "green_bold"))
 
+    #Sort to print alphabetically ordered list of processes to ease reading
```
Comment (tiagofilipe12) [inline flowcraft/generator/process_details.py]: 👍

File: flowcraft/generator/templates/maxbin2.nf [hunk 10]
```diff
@@ -82,14 +86,14 @@ process report_maxbin2_{{ pid }}{
 OUT_binned = Channel.create()
 OUT_unbinned = Channel.create()
 
-chanA = Channel.create()
-chanB = Channel.create()
+failedBinning = Channel.create()
+successfulBinning = Channel.create()
 
-binCh_{{ pid }}.choice(chanA, chanB){ it -> it[3].text == "false\n" ? 0 : 1 }
+binCh_{{ pid }}.choice(failedBinning, successfulBinning){ it -> it[3].text == "false\n" ? 0 : 1 }
```
Comment (tiagofilipe12) [inline flowcraft/generator/templates/maxbin2.nf]: Excellent! Now it is easier to read. 👍

File: docs/user/components/maxbin2.rst (hunk 8 above)
Comment (cimendes) [reply docs/user/components/maxbin2.rst]: I check some examples and I saw both... I went to the more descriptive one and made a mental note (now written here) to change them to that.

File: changelog.md [hunk 11]
```diff
@@ -2,21 +2,33 @@
 
 ## Changes in upcoming release (`dev` branch) 
 
+
+### New features
+
+- Added support for the report system to:
+    - `maxibin2`
```
Comment (ODiogoSilva) [inline changelog.md]: maxibin2? or maxbin2?

File: changelog.md [hunk 12]
```diff
@@ -2,21 +2,33 @@
...
 - Fix parsing of .nextflow.log file when searching for `nextflow run` command.
 
-
 ### Minor/Other changes
 
 - Added option to `dengue_typing` to retrive closest referece sequence and link it 
 with a secondary channel into `mafft`
 - New version of DEN-IM recipe
+- Now prints an ordered list of components
+- moved taxonomy results from `results/annotation/` to `results/taxonomy/`
```
Comment (ODiogoSilva) [inline changelog.md]: Upper case 

```suggestion
- Moved taxonomy results from `results/annotation/` to `results/taxonomy/`
```

File: flowcraft/generator/components/metagenomics.py [hunk 13]
```diff
@@ -41,6 +41,42 @@ def __init__(self, **kwargs):
...
+                - ``input_type``: fastq
+                - ``output_type``: txt
+                - ``ptype``: taxonomic classification
+    """
+    def __init__(self, **kwargs):
+
+        super().__init__(**kwargs)
+
+        self.input_type = "fastq"
+        self.output_type = "txt"
```
Comment (ODiogoSilva) [inline flowcraft/generator/components/metagenomics.py]: The kraken nextflow process doesn't have a mai ``{{output_channel}}``, so it shouldn't have an output_type:

```suggestion
        self.output_type = None
```

File: flowcraft/generator/components/metagenomics.py [hunk 14]
```diff
@@ -41,6 +41,42 @@ def __init__(self, **kwargs):
...
+
+        self.params = {
+            "kraken2DB": {
+                "default": "'minikraken2_v1_8GB'",
+                "description": "Specifies kraken2 database."
```
Comment (ODiogoSilva) [inline flowcraft/generator/components/metagenomics.py]: Do you need to specify a full path? relative path?

File: flowcraft/generator/templates/maxbin2.nf [hunk 15]
```diff
@@ -50,9 +60,45 @@ process maxbin2_{{ pid }} {
...
+OUT_unbinned = Channel.create()
+
+failedBinning = Channel.create()
+successfulBinning = Channel.create()
+
+binCh_{{ pid }}.choice(failedBinning, successfulBinning){ it -> it[3].text == "false\n" ? 0 : 1 }
+
+failedBinning.map{ it -> [it[0], it[1]] }.into(OUT_unbinned)
+
+successfulBinning.map{ it -> [it[2].toString().tokenize('/').last().tokenize('.')[0..-2].join('.'), it[2]]}
```
Comment (ODiogoSilva) [inline flowcraft/generator/templates/maxbin2.nf]: Nice work here :tada:

File: flowcraft/generator/templates/megahit.nf [hunk 16]
```diff
@@ -32,38 +32,39 @@ process megahit_{{ pid }} {
     template "megahit.py"
 
 }
-IN_fastg = Channel.create()
 
 fastg = params.fastg{{ param_id }} ? "true" : "false"
-if (fastg) {
-    process megahit_fastg_{{ pid }}{
+process megahit_fastg_{{ pid }}{
```
Comment (ODiogoSilva) [inline flowcraft/generator/templates/megahit.nf]: Why did you remove the existing conditional that would executed the fastg stuff only when the param was provided?

File: flowcraft/templates/process_tsv.py [hunk 17]
```diff
@@ -0,0 +1,59 @@
+#!/usr/bin/env python3
+
+"""
+Purpose
+-------
+This module is intended to process the output in tsv
+ to generate a report in json format.
+Expected input
```
Comment (ODiogoSilva) [inline flowcraft/templates/process_tsv.py]: Space above the headers here, and the in the others

File: flowcraft/templates/process_tsv.py [hunk 18]
```diff
@@ -0,0 +1,59 @@
...
+--------------
+The following variables are expected whether using NextFlow or the
+:py:func:`main` executor.
+- ``sample_id`` : Sample Identification string.
+- ``tsv``: tsv output.
+Generated output
+----------------
+- ``.report.json``: Data structure for the report
```
Comment (ODiogoSilva) [inline flowcraft/templates/process_tsv.py]: dotfiles (.report.json, .versions, .status, etc) should not count as generated output. This is reserved to non-hidden files.

File: flowcraft/templates/process_tsv.py [hunk 19]
```diff
@@ -0,0 +1,59 @@
...
+
+    report_json = {
+        "tsvData": [{
+            "sample": sample_id,
+            "data": {}
+        }]
+    }
+
+    #web-app excepts a list with all the values in the table.
+    report_json["tsvData"][0]["data"]["MaxBin2"] = list(csv.reader(open(tsv_file), delimiter='\t'))
```
Comment (ODiogoSilva) [inline flowcraft/templates/process_tsv.py]: Why didn't you set the `"MaxBin2"` directly on the literal dict definition. 

``` 
     report_json = {
         "tsvData": [{
             "sample": sample_id,
             "data": {
                 "MaxBin2": list(csv.reader(open(tsv_file), delimiter='\t'))
             }
         }]
     }
```

File: changelog.md (hunk 11 above)
Comment (cimendes) [reply changelog.md]: Captain Diogo always checks the changelog... My bad X) Fixed

File: flowcraft/generator/components/metagenomics.py (hunk 14 above)
Comment (cimendes) [reply flowcraft/generator/components/metagenomics.py]: Full path. I've clarified the param description.

File: flowcraft/templates/process_tsv.py [hunk 20]
```diff
@@ -0,0 +1,59 @@
...
+    SAMPLE_ID = '$sample_id'
+    FILE = '$tsv'
+    logger.debug("Running {} with parameters:".format(
+        os.path.basename(__file__)))
+    logger.debug("SAMPLE_ID: {}".format(SAMPLE_ID))
+    logger.debug("FILE: {}".format(FILE))
+
+@MainWrapper
+def main(sample_id, tsv_file):
+
```
Comment (tiagofilipe12) [inline flowcraft/templates/process_tsv.py]:
```suggestion

# TODO this tsvData could be a single object since it only has one element
```

File: flowcraft/generator/components/metagenomics.py (hunk 14 above)
Comment (ODiogoSilva) [reply flowcraft/generator/components/metagenomics.py]: If so, why is the default not a full path? I'm assuming it's on the image? If yes, then add that information as well.

File: flowcraft/generator/templates/megahit.nf (hunk 16 above)
Comment (cimendes) [reply flowcraft/generator/templates/megahit.nf]: When I tested while I was implementing kraken2, it wasn't working. The process was being executed regardless of the value of the conditional. So I changed it.

File: flowcraft/templates/process_tsv.py (hunk 18 above)
Comment (cimendes) [reply flowcraft/templates/process_tsv.py]: So the generated output is none?

File: flowcraft/templates/process_tsv.py (hunk 19 above)
Comment (cimendes) [reply flowcraft/templates/process_tsv.py]: To ease template extension to other processes. This way it's clear where the template needs to be edited.

File: flowcraft/templates/process_tsv.py (hunk 18 above)
Comment (cimendes) [reply flowcraft/templates/process_tsv.py]: removed this section

File: flowcraft/generator/components/metagenomics.py [hunk 21]
```diff
@@ -41,6 +41,43 @@ def __init__(self, **kwargs):
...
+
+        super().__init__(**kwargs)
+
+        self.input_type = "fastq"
+        self.output_type = None
+
+        self.params = {
+            "kraken2DB": {
+                "default": "'minikraken2_v1_8GB'",
+                "description": "Specifies kraken2 database. Requires full path if database not on"
```
Comment (ODiogoSilva) [inline flowcraft/generator/components/metagenomics.py]:
```suggestion
                "description": "Specifies kraken2 database. Requires full path if database not on "
```

File: flowcraft/generator/templates/megahit.nf (hunk 16 above)
Comment (ODiogoSilva) [reply flowcraft/generator/templates/megahit.nf]: So, in this way both options work fine?

File: flowcraft/templates/process_tsv.py (hunk 19 above)
Comment (ODiogoSilva) [reply flowcraft/templates/process_tsv.py]: I'm not sure how this is clearer or easier to read than the literal definition?

File: flowcraft/generator/templates/megahit.nf (hunk 16 above)
Comment (cimendes) [reply flowcraft/generator/templates/megahit.nf]: yes. I've tested with both options

File: flowcraft/templates/process_tsv.py (hunk 19 above)
Comment (cimendes) [reply flowcraft/templates/process_tsv.py]: I prefer to make the literal definition general and populate the dictionary as needed. This way if the this template needs to be extended for other modules, the line that needs to be substituted by the appropriate method is separated. This has no impact on the execution time, but if you really dislike this, I can alter it.
